torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,540 @@
+    if base1 is None or base2 is None:
+        return False
+    # Check that "args..." are the same too.
+    assert len(view1.args) == len(view2.args)
+    assert len(view1.kwargs) == len(view2.kwargs)
+    # If the two views were created through the same view op (with the same args),
+    # then recursively check if their bases are also matching views.
+    if view1.target is view2.target:
+        if all(x == y for (x, y) in zip(
+               itertools.chain(view2.args[1:], view2.kwargs),
+               itertools.chain(view1.args[1:], view1.kwargs))):
+            return _are_matching_view_tensors(base1, base2)
+    return False
+
+
+
+def _satisfies_inverse_view(view_inverse_node: torch.fx.node.Node, tensor_aliases: Set[torch.fx.node.Node]):
+    for alias in tensor_aliases:
+        # If the current node is a *_scatter op, first check if any of the aliases
+        # correspond to the inverse (e.g. slice_scatter -> slice)
+        if view_inverse_node.target == _VIEW_INVERSE_MAP.get(alias.target, None):
+            # Every view_inverse op is of the form *_scatter(base, mutated_view, args...)
+            # First, check that "base" in the inverse call matches the base in the view op call.
+            if _are_matching_view_tensors(view_inverse_node.args[0], alias.args[0]):
+                # Then, check that "args..." are the same too.
+                # view inverse nodes like *_scatter are always of the form:
+                #     select(base, args...)
+                #     select_scatter(base, mutated_view, args...)
+                assert len(alias.args) + 1 == len(view_inverse_node.args)
+                assert len(alias.kwargs) == len(view_inverse_node.kwargs)
+                if all(x == y for (x, y) in zip(
+                       itertools.chain(view_inverse_node.args[2:], view_inverse_node.kwargs.items()),
+                       itertools.chain(alias.args[1:], alias.kwargs.items()))):
+                    return True
+    return False
+
+
+# This function tests where a tensor's memory is allowed to be re-inplaced.
+# e.g. given:
+#    a.add(b)
+# It returns true if we can convert it into
+#    a.add_(b)
+# We figure this out as follows:
+#   For every alias of a, "x",
+#   EITHER
+#       "x" is *never* used later in the program
+#   OR
+#       there is some existing alias of a, "alias", where
+#       "alias" is used in some slicing view operation like _ = torch.select(alias, args...)
+#       AND "x" is used in the corresponding "scatter" variant of that operator,
+#       e.g. _ = torch.select_scatter(x, mutated_view, args...)
+#       AND the arguments to the view and its scatter operator are the same (args... is equal).
+#   This is effectively undoing the view-inverse logic performed by functionalization.
+def _get_all_later_node_usages(curr_node: torch.fx.node.Node, tensor_aliases: Set[torch.fx.node.Node], op_index: int, *, only_include_view_inverse_nodes: bool = False) -> Set[torch.fx.node.Node]:
```

Comment:
So if this function tests if the tensor can be reinplaced, why is the function named get all later node usages

Reply:
It's also not clear what all of the arguments mean

Reply:
This is not just node usages, it is usages of any alias of the node!

Reply:
Yes - agreed that was pretty unclear, will beef up the description

Reply:
Basically, originally I wanted a function that was just `can_reinplace(...) -> bool`. That function needs to find any candidate `*_scatter*` ops later in the graph that can be removed, in order to figure out if the current node can be replaced. Later on though we need to get any `*_scatter` nodes directly if they exist so that we can remove them from the graph, so I figured I'd have this function directly return those nodes.

So this function really has two modes:
- `only_include_view_inverse_nodes =False`: get all later node usages (and their aliases) of the current self_arg (`curr_node`) that occur later than `op_index` in the graph (where `op_index` is the node we're trying to re-inplace)
- `only_include_view_inverse_nodes` same as the above, but only include the `{view}_scatter` ops